File: unleash/src/main/java/com/silvercar/unleash/strategy/Strategy.java
package com.silvercar.unleash.strategy;

import java.util.List;
import java.util.Map;

import com.silvercar.unleash.Constraint;
import com.silvercar.unleash.UnleashContext;

public interface Strategy {
    String getName();

    boolean isEnabled(Map<String, String> parameters);

    default boolean isEnabled(Map<String, String> parameters, UnleashContext unleashContext) {
        return isEnabled(parameters);
    }

    default boolean isEnabled(Map<String, String> parameters, UnleashContext unleashContext, List<Constraint> constraints) {
        return ConstraintUtil.validate(constraints, unleashContext) && isEnabled(parameters, unleashContext);
    }
}

File: unleash/src/main/java/com/silvercar/unleash/strategy/Strategy.kt
package com.silvercar.unleash.strategy

import com.silvercar.unleash.UnleashContext

interface Strategy {
  val name: String
  fun isEnabled(parameters: Map<String, String>): Boolean
  fun isEnabled(
    parameters: Map<String, String>,
    unleashContext: UnleashContext
  ): Boolean {
    return isEnabled(parameters)
  }

  companion object {
    const val ROLLOUT = "rollout"
    const val PERCENTAGE = "percentage"
    const val GROUP_ID = "groupId"
  }
}
